File: drills/elixir/solutions/nc85_permutations__insert_everywhere.exs
defmodule Solution do
  # Build up instead of choosing: every permutation of n elements is a
  # permutation of n-1 with the new element wedged into one of its n positions.
  # No recursion into a shrinking remainder, and it explains the factorial
  # directly -- one more choice of position at every step.
  def permute(nums) do
    Enum.reduce(nums, [[]], fn value, permutations ->
      for permutation <- permutations,
          at <- 0..length(permutation)//1,
          do: List.insert_at(permutation, at, value)
    end)
  end
end

File: drills/python/solutions/nc85_permutations__insert_everywhere.py
def permute(nums):
    # Build up instead of choosing: every permutation of n elements is a
    # permutation of n-1 with the new element wedged into one of its n
    # positions. No recursion into a shrinking remainder, and it explains the
    # factorial directly -- one more choice of position at every step.
    permutations = [[]]
    for value in nums:
        permutations = [
            permutation[:at] + [value] + permutation[at:]
            for permutation in permutations
            for at in range(len(permutation) + 1)
        ]
    return permutations
